File: demo/com/natpryce/snodge/demo/JsonEventFormatSnodgeTest.java
package com.natpryce.snodge.demo;

import com.google.gson.JsonPrimitive;
import com.natpryce.snodge.JsonMutator;
import com.natpryce.snodge.Mutagen;
import com.natpryce.snodge.mutagens.AddObjectProperty;
import com.natpryce.snodge.mutagens.ReorderObjectProperties;
import org.junit.Test;

import java.io.IOException;
import java.net.URI;

import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.junit.MatcherAssert.assertThat;

public class JsonEventFormatSnodgeTest {
    static final int mutationCount = 100;

//    JsonEventFormat format = new JsonEventFormat();
    // Replace the line below with the line above to see the test catch failures
    RobustJsonEventFormat format = new RobustJsonEventFormat();

    ServiceEvent originalEvent = new ServiceEvent(
            System.currentTimeMillis(),
            URI.create("iotp://192.168.1.47/sensor/temp"),
            ServiceEvent.Type.READY);

    @Test
    public void parsesEventSuccessfullyOrThrowsIOException() throws IOException {
        new JsonMutator().forStrings().mutate(format.serialise(originalEvent), mutationCount)
                .forEach(this::assertParsesEventSuccessfullyOrThrowsIOException);
    }

    @Test
    public void jsonPropertiesCanOccurInAnyOrder() throws IOException {
        assertSerialisationUnaffectedBy(new ReorderObjectProperties());
    }

    @Test
    public void ignoresUnrecognisedProperties() throws IOException {
        assertSerialisationUnaffectedBy(new AddObjectProperty(new JsonPrimitive("new-property-value")));
    }

    private void assertSerialisationUnaffectedBy(Mutagen mutagen) throws IOException {
        new JsonMutator(mutagen).forStrings().mutate(format.serialise(originalEvent), mutationCount)
                .forEach(this::assertParsedEventEqualToOriginal);
    }

    private void assertParsedEventEqualToOriginal(String json) {
        try {
            assertThat("deserialised JSON: " + json, format.deserialise(json), equalTo(originalEvent));
        }
        catch (Exception e) {
            fail("deserialisation failed for JSON: " + json, e);
        }
    }

    private void assertParsesEventSuccessfullyOrThrowsIOException(String json) {
        try {
            format.deserialise(json);
        }
        catch (IOException e) {
            // allowed
        }
        catch (Exception e) {
            fail("unexpected exception for JSON input: " + json, e);
        }
    }

    private void fail(String detailMessage, Exception e) {
        AssertionError failure = new AssertionError(detailMessage);
        failure.initCause(e);
        throw failure;
    }
}

File: demo/com/natpryce/snodge/demo/JsonEventFormatSnodgeTest.kt
package com.natpryce.snodge.demo

import com.natpryce.hamkrest.assertion.assertThat
import com.natpryce.hamkrest.equalTo
import com.natpryce.jsonk.JsonElement
import com.natpryce.jsonk.JsonString
import com.natpryce.snodge.Mutagen
import com.natpryce.snodge.json.addObjectProperty
import com.natpryce.snodge.json.defaultJsonMutagens
import com.natpryce.snodge.json.forStrings
import com.natpryce.snodge.json.reorderObjectProperties
import com.natpryce.snodge.mutants
import org.junit.Test
import java.io.IOException
import java.net.URI
import java.util.Random

class JsonEventFormatSnodgeTest {
    
    // uncomment one of the formats below to see the test detect defects
//    val format = ShonkyJsonEventFormat();
//    val format = ReflectomagicJsonEventFormat()
    val format = RobustJsonEventFormat()
    
    internal val originalEvent = ServiceEvent(
        System.currentTimeMillis(),
        URI.create("iotp://192.168.1.47/sensor/temp"),
        FAILED("testing"))
    
    @Test
    fun parsesEventSuccessfullyOrThrowsIOException() {
        Random().mutants(defaultJsonMutagens().forStrings(), mutationCount, format.serialise(originalEvent)).forEach { s ->
            assertParsesEventSuccessfullyOrThrowsIOException(s)
        }
    }
    
    @Test
    fun jsonPropertiesCanOccurInAnyOrder() {
        assertSerialisationUnaffectedBy(reorderObjectProperties())
    }
    
    @Test
    fun ignoresUnrecognisedProperties() {
        assertSerialisationUnaffectedBy(addObjectProperty(JsonString("new-property-value")))
    }
    
    private fun assertSerialisationUnaffectedBy(mutagen: Mutagen<JsonElement>) {
        Random().mutants(mutagen.forStrings(), 1000, format.serialise(originalEvent))
            .forEach { assertParsedEventEqualToOriginal(it) }
    }
    
    private fun assertParsedEventEqualToOriginal(json: String) {
        try {
            assertThat("deserialised JSON: " + json, format.deserialise(json), equalTo(originalEvent))
        }
        catch (e: Exception) {
            fail("deserialisation failed for JSON: " + json, e)
        }
    }
    
    private fun assertParsesEventSuccessfullyOrThrowsIOException(json: String) {
        try {
            format.deserialise(json)
        }
        catch (e: IOException) {
            // allowed
        }
        catch (e: Exception) {
            fail("unexpected exception for JSON input: " + json, e)
        }
        
    }
    
    private fun fail(detailMessage: String, e: Exception): Nothing = throw AssertionError(detailMessage, e)
    
    internal val mutationCount = 1000
}
